### PR TITLE
Fix: Logic for demotion rules around scalar ad types and data matrices

### LINKED ISSUE #1232: SoA optimization fail?
Using `--O1` the following code fails to use SoA for `beta` and `mu`
```
data {
  int N;
}
parameters {
  real alpha;
  vector[N] beta;
}
model{
  vector[N] mu = alpha + rep_vector(0.0, N) + beta;
  target += sum(mu);
}
```
but the following equivalent code does work (showing only the model block)
```
model{
  vector[N] mu = alpha + (rep_vector(0.0, N) + beta);
  target += sum(mu);
}
```
but it's not just the order of summing as following works, too
```
model{
  vector[N] tmp = rep_vector(0.0, N);
  vector[N] mu = alpha + tmp + beta;
  target += sum(mu);
}
```

@WardBrian commented
>I believe the problem is the formulation `alpha + rep_vector(0.0, N) + beta` leads to the first being a `Plus((AutoDiffable UReal) (DataOnly UVector))`, which does not contain any eigen autodiff variables so the optimization says oh, SoA can not be used

> whereas `alpha + (rep_vector(0.0, N) + beta)` is a `Plus((DataOnly UVector) (AutoDiffable UVector))` which does contain it

> `tmp` is a model-block level variable which means it is always a AutoDiffable

Is there anything to be done to make the first code to use SoA, as figuring out these subtleties can be challenging for users

### PR DESCRIPTION
**Pull request by SteveBronder**
### Summary

Fixes #1232 where some mixes of scalars, data matrices, and autodiff matrices would not be promoted to SoA matrices appropriately. 

For the following code we would not promote to SoA when we should have 

```stan
data {
  int N;
}
parameters {
  real alpha;
  vector[N] beta;
}
model{
  vector[N] mu = alpha + rep_vector(0.0, N) + beta;
  target += sum(mu);
}
```

but for the following model block it does make SoA matrices correctly

```stan
model{
  vector[N] mu = alpha + (rep_vector(0.0, N) + beta);
  target += sum(mu);
}
```

This is because of a bug in the logic for detecting whether we should demote expressions when we have a mix of scalars and data matrices. 

The difference between the two model blocks above is that, in pseudo mir code, the first has a scalar and data matrix operation inside of the expression while the second, because of the parenthesis, has a scalar and autodiff matrix operation

```
assign(mu, add(add(alpha, rep_vector(0.0, N)), beta)
```

```
assign(mu, add(alpha, add(rep_vector(0.0, N), beta))
```

The first fails because the optimizer sees a scalar and data matrix operation and fails the whole line. But the second just has a scalar and autodiff matrix operation so it does not fail.

Though we do want the following to fail for trying to promote `p`. In this code, `p` is derived from an autodiff scalar and a data matrix, which we can't promote to SoA.

```stan
data {
  int J;
  array[J] int n;
  vector[J] x;
  array[J] int y;
  real r;
  real R;
}
transformed data {
  vector[J] threshold_angle = asin((R - r) ./ x);
}
parameters {
  real<lower=0> sigma;
}
model {
  vector[J] p = 2 * Phi(threshold_angle / sigma) - 1;
  y ~ binomial(n, p);
}
```

I found a bug in the logic where if the optimizer saw any `op(autodiff scalar, data matrix)` operations before it saw a `autodiff matrix` it would fail the expression and the left hand side would need to be demoted to a AoS.

This PR fixes that by, for each expression with a `autodiff matrix` type, it goes through checking if it is derived only by autodiff scalars and data matrices. If so then we do a second pass to check if there are any autodiff matrices in the expression graph for that statement that are derived outside of the expression. If we find an autodiff matrix that is not derived in the expression then we do not force demotion of the left hand side.

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [ ] If a user-facing facing change was made, the documentation PR is here: <LINK>
    - [x] OR, no user-facing changes were made

## Release notes

Fix logic for deducing whether mixes of scalars, data matrices, and autodiff matrices can be promoted to SoA

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: src/analysis_and_optimization/Memory_patterns.ml [hunk 1]
```diff
@@ -195,111 +195,60 @@ and query_initial_demotable_funs (in_loop : bool) (acc : string Set.Poly.t)
...
-     |TupleProjection (expr, _) ->
-        is_any_soa_supported_expr expr
-    | Var (_ : string) | Lit ((_ : Expr.Fixed.Pattern.litType), (_ : string)) ->
-        true
-    | TernaryIf (_, texpr, fexpr) ->
-        is_any_soa_supported_expr texpr && is_any_soa_supported_expr fexpr
-    | EAnd (lhs, rhs) | EOr (lhs, rhs) ->
-        is_any_soa_supported_expr lhs && is_any_soa_supported_expr rhs
+let print_string s = print_endline s
+let print_set s = Set.Poly.iter ~f:print_string s
```
Comment (WardBrian) [inline src/analysis_and_optimization/Memory_patterns.ml:199]:
```suggestion

```

File: src/analysis_and_optimization/Memory_patterns.ml [hunk 2]
```diff
@@ -195,111 +195,60 @@ and query_initial_demotable_funs (in_loop : bool) (acc : string Set.Poly.t)
...
-  | UserDefined ((_ : string), (_ : bool Fun_kind.suffix)) -> false
-  | CompilerInternal (_ : 'a Internal_fun.t) -> true
-  | Fun_kind.StanLib (name, (_ : bool Fun_kind.suffix), _) -> (
-    match name with
-    | "check_matching_dims" -> true
-    | _ ->
-        is_fun_soa_supported name exprs
-        && List.exists ~f:is_any_soa_supported_expr exprs )
+let print_bool nn bb =
+  Printf.printf "%s: %s \n" nn (if bb then "true" else "false")
```
Comment (WardBrian) [inline src/analysis_and_optimization/Memory_patterns.ml:202]:
```suggestion

```

File: src/middle/Stan_math_signatures.ml [hunk 3]
```diff
@@ -2166,8 +2166,8 @@ let () =
         (List.range 1 3) )
     bare_types ;
   add_unqualified ("rep_matrix", ReturnType UMatrix, [UReal; UInt; UInt], SoA) ;
-  add_unqualified ("rep_matrix", ReturnType UMatrix, [UVector; UInt], AoS) ;
-  add_unqualified ("rep_matrix", ReturnType UMatrix, [URowVector; UInt], AoS) ;
+  add_unqualified ("rep_matrix", ReturnType UMatrix, [UVector; UInt], SoA) ;
+  add_unqualified ("rep_matrix", ReturnType UMatrix, [URowVector; UInt], SoA) ;
```
Comment (WardBrian) [inline src/middle/Stan_math_signatures.ml:2170]: Mind updating https://github.com/stan-dev/stanc3/blob/master/test/integration/good/function-signatures/math/matrix/rep_matrix.stan?

File: src/analysis_and_optimization/Memory_patterns.ml [hunk 4]
```diff
@@ -195,111 +195,60 @@ and query_initial_demotable_funs (in_loop : bool) (acc : string Set.Poly.t)
...
 
-(**
-  Return true if the rhs expression of an assignment contains only
-   combinations of AutoDiffable Reals and Data Matrices
- *)
-let rec is_any_ad_real_data_matrix_expr
-    Expr.Fixed.{pattern; meta= Expr.Typed.Meta.{adlevel; _}} : bool =
-  if UnsizedType.is_dataonlytype adlevel then false
-  else
+let rec get_lowest_level
```
Comment (WardBrian) [inline src/analysis_and_optimization/Memory_patterns.ml:204]: I think these functions deserve a docstring (and possibly a better name?)

File: src/analysis_and_optimization/Memory_patterns.ml [hunk 5]
```diff
@@ -336,57 +285,65 @@ let rec query_initial_demotable_stmt (in_loop : bool) (acc : string Set.Poly.t)
...
-              (Set.Poly.union assign_demotes (query_var_eigen_names rhs))
+              (Set.Poly.union rhs_and_idx_demotions (query_var_eigen_names rhs))
+              name
+        | _ -> rhs_and_idx_demotions in
+      let assign_demotions =
+        let is_eigen_stmt = UnsizedType.contains_eigen_type rhs.meta.type_ in
+        if is_eigen_stmt then
+          (* LHS (2)*)
+          let is_rhs_not_promoteable_to_soa =
+            let blah = get_lowest_level rhs in
```
Comment (WardBrian) [inline src/analysis_and_optimization/Memory_patterns.ml:317]:
```suggestion
            let expression_parts = get_lowest_level rhs in
```

Or any other name! Just not blah

**Review by WardBrian — CHANGES_REQUESTED** [review]
A few (hopefully quick) comments. The overall logic looks good to me

File: src/analysis_and_optimization/Memory_patterns.ml (hunk 5 above)
Comment (SteveBronder) [reply src/analysis_and_optimization/Memory_patterns.ml:317]: lmao sorry that was from debugging

File: src/analysis_and_optimization/Memory_patterns.ml (hunk 4 above)
Comment (SteveBronder) [reply src/analysis_and_optimization/Memory_patterns.ml]: Just added doc to explain what the recursion is looking for. And gave these better names

**Comment by SteveBronder** [comment]
Think this is ready for review!

**Review by WardBrian — APPROVED** [review]
LGTM (assuming tests pass)! Thanks for taking this on

**Pull request merged**